### PR TITLE
python3Packages.matplotlib: 3.3.4 -> 3.4.1

### PR DESCRIPTION
**Pull request by veprbl**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by veprbl** [comment]
python2 tests don't work because of some pytest/pluggy incompatibility
```diff
diff --git a/pkgs/development/python-modules/matplotlib/2.nix b/pkgs/development/python-modules/matplotlib/2.nix
index e0879050704..c910a655558 100644
--- a/pkgs/development/python-modules/matplotlib/2.nix
+++ b/pkgs/development/python-modules/matplotlib/2.nix
@@ -1,4 +1,6 @@
-{ lib, stdenv, fetchPypi, writeText, python, buildPythonPackage, pycairo, backports_functools_lru_cache
+{ lib, stdenv, fetchPypi, fetchurl, writeText, python
+, buildPythonPackage, pycairo, pytestCheckHook
+, backports_functools_lru_cache
 , which, cycler, dateutil, nose, numpy, pyparsing, sphinx, tornado, kiwisolver
 , freetype, libpng, pkg-config, mock, pytz, pygobject3, gobject-introspection, functools32, subprocess32
 , fetchpatch
@@ -8,7 +10,9 @@
 , enableTk ? !stdenv.isDarwin, tcl ? null, tk ? null, tkinter ? null, libX11 ? null
 , enableQt ? false, pyqt4
 , Cocoa
+, enableTests ? false
 , pythonOlder
+, matplotlib
 }:
 
 assert enableGhostscript -> ghostscript != null;
@@ -46,8 +50,9 @@ buildPythonPackage rec {
     ++ lib.optional stdenv.isDarwin [ Cocoa ];
 
   propagatedBuildInputs =
-    [ cycler dateutil nose numpy pyparsing tornado freetype kiwisolver
-      libpng mock pytz ]
+    [ cycler dateutil nose numpy pyparsing tornado kiwisolver libpng
+      mock pytz ]
+    ++ lib.optionals (!enableTests) [ freetype ]
     ++ lib.optional (pythonOlder "3.3") backports_functools_lru_cache
     ++ lib.optionals enableGtk3 [ cairo pycairo gtk3 gobject-introspection pygobject3 ]
     ++ lib.optionals enableTk [ tcl tk tkinter libX11 ]
@@ -56,10 +61,29 @@ buildPythonPackage rec {
 
   passthru.config = {
     directories = { basedirlist = "."; };
+    test = {
+      local_freetype = enableTests;
+    };
+    packages = {
+      tests = enableTests;
+    };
   };
   setup_cfg = writeText "setup.cfg" (lib.generators.toINI {} passthru.config);
+
+  # running tests requires a specific version of freetype
+  freetype_src =
+    if enableTests then
+      fetchurl {
+        url = "mirror://sourceforge/project/freetype/freetype2/2.6.1/freetype-2.6.1.tar.gz";
+        sha256 = "053h6qn17r2rav3wisvipyxvlylqdplyhcljwbyfi8bdvbxpsg0a";
+      }
+    else null;
+
   preBuild = ''
     cp "$setup_cfg" ./setup.cfg
+  '' + lib.optionalString enableTests ''
+    mkdir build/
+    cp "$freetype_src" build/"$(stripHash "$freetype_src")"
   '';
 
   # Matplotlib tries to find Tcl/Tk by opening a Tk window and asking the
@@ -75,10 +99,14 @@ buildPythonPackage rec {
     lib.optionalString enableTk
       "sed -i '/self.tcl_tk_cache = None/s|None|${tcl_tk_cache}|' setupext.py";
 
-  # Matplotlib needs to be built against a specific version of freetype in
-  # order for all of the tests to pass.
-  doCheck = false;
+  doCheck = enableTests;
+  checkInputs = [ pytestCheckHook ];
+
+  preCheck = ''
+    cd "$out"
+  '';
 
+  passthru.tests = [ (matplotlib.override { enableTests = true; }) ];
   meta = with lib; {
     description = "Python plotting library, making publication quality plots";
     homepage    = "https://matplotlib.org/";
```

**Comment by veprbl** [comment]
@GrahamcOfBorg python3Packages.matplotlib python3Packages.matplotlib.tests
@GrahamcOfBorg python2Packages.matplotlib

**Comment by veprbl** [comment]
@GrahamcOfBorg build python3Packages.matplotlib python3Packages.matplotlib.tests
@GrahamcOfBorg build python2Packages.matplotlib

**Comment by veprbl** [comment]
@GrahamcOfBorg eval

**Comment by veprbl** [comment]
@GrahamcOfBorg build python3Packages.matplotlib python3Packages.matplotlib.tests
@GrahamcOfBorg build python2Packages.matplotlib

**Comment by veprbl** [comment]
@GrahamcOfBorg build python3Packages.matplotlib python3Packages.matplotlib.tests
@GrahamcOfBorg build python2Packages.matplotlib

File: pkgs/development/python-modules/matplotlib/default.nix [hunk 1]
```diff
@@ -1,14 +1,16 @@
-{ lib, stdenv, fetchPypi, python, buildPythonPackage, isPy3k, pycairo, backports_functools_lru_cache
+{ lib, stdenv, fetchPypi, fetchzip, writeText, python
+, buildPythonPackage, isPy3k, pycairo, pytestCheckHook
 , which, cycler, dateutil, nose, numpy, pyparsing, sphinx, tornado, kiwisolver
-, freetype, libpng, pkg-config, mock, pytz, pygobject3, gobject-introspection
+, freetype, qhull, libpng, pkg-config, mock, pytz, pygobject3, gobject-introspection
 , certifi, pillow
 , enableGhostscript ? true, ghostscript ? null, gtk3
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/matplotlib/default.nix:6]:
```suggestion
, enableGhostscript ? true, ghostscript, gtk3
```

File: pkgs/development/python-modules/matplotlib/default.nix [hunk 2]
```diff
@@ -1,14 +1,16 @@
...
 # darwin has its own "MacOSX" backend
 , enableTk ? !stdenv.isDarwin, tcl ? null, tk ? null, tkinter ? null, libX11 ? null
 , enableQt ? false, pyqt5 ? null
+, enableTests ? false
 , Cocoa
-, pythonOlder
+, matplotlib
 }:
 
 assert enableGhostscript -> ghostscript != null;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/matplotlib/default.nix:16]:
```suggestion
```

File: pkgs/development/python-modules/matplotlib/default.nix [hunk 3]
```diff
@@ -1,14 +1,16 @@
...
+, buildPythonPackage, isPy3k, pycairo, pytestCheckHook
 , which, cycler, dateutil, nose, numpy, pyparsing, sphinx, tornado, kiwisolver
-, freetype, libpng, pkg-config, mock, pytz, pygobject3, gobject-introspection
+, freetype, qhull, libpng, pkg-config, mock, pytz, pygobject3, gobject-introspection
 , certifi, pillow
 , enableGhostscript ? true, ghostscript ? null, gtk3
 , enableGtk3 ? false, cairo
 # darwin has its own "MacOSX" backend
 , enableTk ? !stdenv.isDarwin, tcl ? null, tk ? null, tkinter ? null, libX11 ? null
 , enableQt ? false, pyqt5 ? null
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/matplotlib/default.nix:10]:
```suggestion
, enableQt ? false, pyqt5
```

File: pkgs/development/python-modules/matplotlib/default.nix [hunk 4]
```diff
@@ -20,14 +22,14 @@ assert enableTk -> (tcl != null)
 assert enableQt -> pyqt5 != null;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/matplotlib/default.nix:22]:
```suggestion
```

File: pkgs/development/python-modules/matplotlib/default.nix [hunk 5]
```diff
@@ -1,14 +1,16 @@
...
+{ lib, stdenv, fetchPypi, fetchzip, writeText, python
+, buildPythonPackage, isPy3k, pycairo, pytestCheckHook
 , which, cycler, dateutil, nose, numpy, pyparsing, sphinx, tornado, kiwisolver
-, freetype, libpng, pkg-config, mock, pytz, pygobject3, gobject-introspection
+, freetype, qhull, libpng, pkg-config, mock, pytz, pygobject3, gobject-introspection
 , certifi, pillow
 , enableGhostscript ? true, ghostscript ? null, gtk3
 , enableGtk3 ? false, cairo
 # darwin has its own "MacOSX" backend
 , enableTk ? !stdenv.isDarwin, tcl ? null, tk ? null, tkinter ? null, libX11 ? null
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/matplotlib/default.nix:9]:
```suggestion
, enableTk ? !stdenv.isDarwin, tcl, tk , tkinter , libX11
```

**Review by SuperSandro2000 — COMMENTED** [review]
Those ? null only hide missing packages as they can't actually be ever null.

**Comment by veprbl** [comment]
@GrahamcOfBorg build python3Packages.matplotlib
@GrahamcOfBorg build python2Packages.matplotlib

**Comment by FRidh** [comment]
> Those ? null only hide missing packages as they can't actually be ever null.

Why can't they ever be null? If I call the expression *without* passing that parameter, it will be null. This needs to be undone, because now, calling the expression when you do not want tk, does require you to pass in tk, even though it will be unused. That's why they have `? null`.

These asserts are also not pointless. They perform an actual test. You may not like them (e.g. because of how they impact overriding) but they are not pointless.

**Comment by veprbl** [comment]
> > Those ? null only hide missing packages as they can't actually be ever null.
> 
> Why can't they ever be null? If I call the expression _without_ passing that parameter, it will be null. This needs to be undone, because now, calling the expression when you do not want tk, does require you to pass in tk or at least something for tk, even though it will be unused. That's why they have `? null`.

No asserts are currently triggered by default:
```
# nix-instantiate -I nixpkgs=channel:nixpkgs-unstable -E 'with import <nixpkgs> {}; python3Packages.matplotlib.override { enableGtk3 = true; enableTk = true; enableQt = true; }'
warning: you did not specify '--add-root'; the result might be removed by the garbage collector
/nix/store/c8cvz2ps9yxv2aifdfzqsygb3a7qvcaw-python3.8-matplotlib-3.3.4.drv
```
This would only cover a hypothetical situation where one tries to deliberately pass `null` values.

> These asserts are also not pointless. They perform an actual test. You may not like them (e.g. because of how they impact overriding) but they are not pointless.

IMO they are not just pointless, but also harmful. Remember #53487

**Comment by veprbl** [comment]
If there are no immediate concerns, let's merge. I don't mind if we revert the change that removes `assert`s.

**Comment by SuperSandro2000** [comment]
> Why can't they ever be null?

If you are not deliberately passing in null for those inputs they are never triggered in nixpkgs and do not add any real value.

If are disabling tk like ``python3Packages.matplotlib.override { enableGtk3 = false; enableTk = false; enableQt = false; }`` it works perfectly fine in nixpkgs.

Those asserts are being removed since months from many not core packages and are not allowed for new packages because of the above mentioned issues.

**Pull request merged**